File: src/moneyTracker.js
function Moneytracker() {
  this._transactions = [];
  this._totalBalance = 0;
};

Moneytracker.prototype.recordEarning = function (money, date) {
  this._transactions.push(new Transaction(money, date))
};

Moneytracker.prototype.recordExpenditure = function (money, date) {
  this._transactions.push(new Transaction(money, date))
};

Moneytracker.prototype._updateTotalBalance = function () {
  for(var i = 0; i < this._transactions.length; i++) {
    this._totalBalance += this._transactions[i].credit + this._transactions[i].debit;
    this._transactions[i].balance = this._totalBalance;
  };
};

Moneytracker.prototype.printStatement = function () {
  this._updateTotalBalance();
  var statement = new Statement(this._transactions);
  return statement.createStatement;
};

File: lib/money_tracker.rb
require_relative 'transactions.rb'
require 'date'

class MoneyTracker

  attr_reader :balance, :transactions

  def initialize
    @balance = 0
    @transactions = Transactions.new
  end

  def earn(amount)
    @balance += amount
    save_transaction(amount, balance)
  end

  def spend(amount)
    @balance -= amount
    save_transaction(-amount, balance)
  end

  def print_statement
    @transactions.format_header
    @transactions.format_statement
  end

  private

  def save_transaction(amount, balance)
    @transactions.store_transaction(amount, balance)
  end
end
